File: Dataset/utils/frft.py
import numpy as np
from fconv import fconv  # 假设 fconv 已经实现
from interp import interp  # 假设 interp 已经实现

def frft(f, a):
    """
    快速分数傅立叶变换
    参数:
        f (ndarray): 信号采样（一维数组）
        a (float): 分数傅立叶变换阶次
    返回:
        ndarray: 变换后的信号
    """
    # 将信号转换为一维列向量
    f = np.asarray(f).flatten()
    N = len(f)
    
    # 计算排列索引：MATLAB中 shft = rem((0:N-1)+fix(N/2), N)+1
    shft = ((np.arange(N) + (N // 2)) % N)
    sN = np.sqrt(N)
    a = a % 4

    # 特殊情况处理
    if np.isclose(a, 0):
        return f
    if np.isclose(a, 2):
        return np.flipud(f)
    if np.isclose(a, 1):
        out = np.empty_like(f, dtype=complex)
        out[shft] = np.fft.fft(f[shft]) / sN
        return out
    if np.isclose(a, 3):
        out = np.empty_like(f, dtype=complex)
        out[shft] = np.fft.ifft(f[shft]) * sN
        return out

    # 降低到区间 0.5 < a < 1.5
    if a > 2.0:
        a = a - 2
        f = np.flipud(f)
    if a > 1.5:
        a = a - 1
        f[shft] = np.fft.fft(f[shft]) / sN
    if a < 0.5:
        a = a + 1
        f[shft] = np.fft.ifft(f[shft]) * sN

    # 一般情况： 0.5 < a < 1.5
    alpha = a * np.pi / 2
    tana2 = np.tan(alpha / 2)
    sina = np.sin(alpha)
    
    # 增加采样率：在信号 f 前后各填充 N-1 个零，同时中间进行 sinc 插值
    f = np.concatenate([np.zeros(N - 1), interp(f), np.zeros(N - 1)])
    
    # chirp 预乘
    r = np.arange(-2 * N + 2, 2 * N - 2 + 1, 2)  # 等价于 MATLAB: [-(2*N-2)+? 实际范围为 -2*N+2 到 2*N-2，步长为2
    chrp = np.exp(-1j * np.pi / N * (tana2 / 4) * (r ** 2))
    f = chrp * f
    
    c = np.pi / N / sina / 4
    # 构造卷积核：索引从 -(4*N-4) 到 4*N-4，步长为1
    r2 = np.arange(-(4 * N - 4), 4 * N - 4 + 1)
    kernel = np.exp(1j * c * (r2 ** 2))
    Faf = fconv(kernel, f)
    
    # MATLAB: Faf = Faf(4*N-3:8*N-7)*sqrt(c/pi);
    # Python 索引转换：起始索引 = (4*N-3) - 1 = 4*N-4, 结束索引为 8*N-7（切片不包含结束位置）
    Faf = Faf[4 * N - 4 : 8 * N - 7] * np.sqrt(c / np.pi)
    
    # chirp 后乘
    Faf = chrp * Faf
    
    # 归一化常数： MATLAB: Faf = exp(-1i*(1-a)*pi/4)*Faf(N:2:end-N+1);
    # Python 中对应的切片：从索引 N-1 到 len(Faf)-(N-1)（不包含终点），步长为2
    Faf = np.exp(-1j * (1 - a) * np.pi / 4) * Faf[N - 1 : len(Faf) - (N - 1) : 2]
    
    return Faf

File: Dataset/utils/frft.m
function Faf = frft(f, a)
% The fast Fractional Fourier Transform
% input: f = samples of the signal信号
%        a = fractional power 分数傅立叶变换的阶次
% output: Faf = fast Fractional Fourier transform
%当一个函数的输入参量的个数超出了规定的范围，MATLAB函数nargchk提供了统一的响应
%输入参量最少是两个最多是两个
error(nargchk(2, 2, nargin));
%将矩阵f的每一列元素堆积起来，成为一个列向量，matlab存储方式
f = f(:);
N = length(f);
shft = rem((0:N-1)+fix(N/2),N)+1;%rem()取余数；fix（）取整体部分;总体是右边的一半移到左边
sN = sqrt(N);
a = mod(a,4);%周期为4
% do special cases
if (a==0), Faf = f; return; end;
if (a==2), Faf = flipud(f); return; end;%flipud turn oppsite实现上下翻转
if (a==1), Faf(shft,1) = fft(f(shft))/sN; return; end 
if (a==3), Faf(shft,1) = ifft(f(shft))*sN; return; end
% reduce to interval 0.5 < a < 1.5
if (a>2.0), a = a-2; f = flipud(f); end
if (a>1.5), a = a-1; f(shft,1) = fft(f(shft))/sN; end
if (a<0.5), a = a+1; f(shft,1) = ifft(f(shft))*sN; end
% the general case for 0.5 < a < 1.5
alpha = a*pi/2;
tana2 = tan(alpha/2);
sina = sin(alpha);
f = [zeros(N-1,1) ; interp(f) ; zeros(N-1,1)];%increase sampling rate
% chirp premultiplication
chrp = exp(-i*pi/N*tana2/4*(-2*N+2:2*N-2)'.^2);
f = chrp.*f;
% chirp convolution
c = pi/N/sina/4;
Faf = fconv(exp(i*c*(-(4*N-4):4*N-4)'.^2),f);
Faf = Faf(4*N-3:8*N-7)*sqrt(c/pi);
% chirp post multiplication
Faf = chrp.*Faf;
% normalizing constant
Faf = exp(-i*(1-a)*pi/4)*Faf(N:2:end-N+1);

function xint=interp(x)
% sinc interpolation
N = length(x);
y = zeros(2*N-1,1);
y(1:2:2*N-1) = x;
xint = fconv(y(1:2*N-1), sinc([-(2*N-3):(2*N-3)]'/2));
xint = xint(2*N-2:end-2*N+3);

function z = fconv(x,y)
% convolution by fft
N = length([x(:);y(:)])-1;
P = 2^nextpow2(N);
z = ifft( fft(x,P) .* fft(y,P));
z = z(1:N); 
